### PR TITLE
Changes to enable Kaleidoscope plugin API v2

### PR DESCRIPTION
**Pull request by noseglasses**
Only small changes were required to adapt the plugin to the Kaleidoscope plugin API v2. I also fixed the example sketches.

### CODE REVIEW

File: examples/fc_example.ino [hunk 1]
```diff
@@ -27,9 +27,6 @@
 // Support for "Numpad" mode, which is mostly just the Numpad specific LED mode
 #include "Kaleidoscope-NumPad.h"
 
-// Support for an "LED off mode"
-#include "LED-Off.h"
```
Comment (noseglasses) [inline examples/fc_example.ino:31]: LED-Off seems not to be supported anymore? Maybe it was renamed?

File: src/Kaleidoscope-LEDEffect-FunctionalColor.cpp [hunk 2]
```diff
@@ -41,8 +41,8 @@ void LEDFunctionalColor::FCPlugin::brightnessDown(uint8_t keyState) {
   LEDFunctionalColor::FCPlugin::lastFC->thisBrightnessDown(keyState);
 }
 
-//setKeyLed accepts a Key position and sets it to the appropriate color from a lookup function
-void LEDFunctionalColor::FCPlugin::setKeyLed(uint8_t r, uint8_t c) {
+//refreshAt accepts a Key position and sets it to the appropriate color from a lookup function
+void LEDFunctionalColor::FCPlugin::refreshAt(byte r, byte c) {
```
Comment (noseglasses) [inline src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:45]: API function was renamed.

File: src/Kaleidoscope-LEDEffect-FunctionalColor.cpp [hunk 3]
```diff
@@ -63,8 +63,7 @@ void LEDFunctionalColor::FCPlugin::refresh(void) {
   // Loop through every row and column and set each LED based on its key's function
   for (uint8_t r = 0; r < ROWS; r++) {
     for (uint8_t c = 0; c < COLS; c++) {
-      Key k = Layer.lookupOnActiveLayer(r, c);
```
Comment (noseglasses) [inline src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:66]: Seems that line does not have any effect. Compiler warns about it.

File: src/keygroups.h [hunk 4]
```diff
@@ -123,7 +123,7 @@ inline constexpr bool isLang(const Key& k) {
 inline constexpr bool isModifier(const Key& k) {
 	return ( k.flags == KEY_FLAGS &&
 		(k.keyCode == (Key_PcApplication).keyCode ||
-		(k.keyCode >= (Key_LeftControl).keyCode) && (k.keyCode <= (Key_RightGui).keyCode)) )?true:false;
+		((k.keyCode >= (Key_LeftControl).keyCode) && (k.keyCode <= (Key_RightGui).keyCode))) )?true:false;
```
Comment (noseglasses) [inline src/keygroups.h:126]: Added a bracket to silence a compiler warning.

File: examples/fc_example.ino (hunk 1 above)
Comment (algernon) [reply examples/fc_example.ino:31]: It's included by default with LEDControl now.

**Comment by jdlien** [comment]
I have created a version1 branch for old firmware versions. I haven't seen anything that appears problematic yet, so I'll merge this request into master.

**Comment by jdlien** [comment]
I tried using the latest version of the Arduino IDE (1.8.8) from arduino.cc and this made no difference.
I've been poking at this for a while and can't figure out why the compiler is complaining about these undefined references. It's almost like it's reading a different file, but the location it mentions is definitely the latest version of FunctionalColor (that is, the one you modified in this PR).

I'm at a loss here... In my test .ino I just took the default included one from Kaleidoscope and added these lines after the initial includes:
```c++
#include <Kaleidoscope-LEDEffect-FunctionalColor.h>
using namespace kaleidoscope::LEDFunctionalColor;
FCPlugin funColor;
```

The complete compiler output follows, in case it reveals something I didn't already mention.
```
WARNING: Category 'Debug' in library Kaleidoscope-Hardware-Virtual is not valid. Setting to 'Uncategorized'
libraries/Kaleidoscope-LEDEffect-FunctionalColor/Kaleidoscope-LEDEffect-FunctionalColor.a(Kaleidoscope-LEDEffect-FunctionalColor.cpp.o): In function `cRGB kaleidoscope::LEDFunctionalColor::groupColorLookup<kaleidoscope::LEDFunctionalColor::colorMapDefault>(kaleidoscope::Key const&, bool&)':
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::control'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
Multiple libraries were found for "HID.h"
 Used: /Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/HID
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
 Not used: /Applications/Arduino.app/Contents/Java/hardware/arduino/avr/libraries/HID
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::gui'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
/Users/jdlien/Library/Arduino15/packages/keyboardio/hardware/avr/1.94.0-beta/libraries/Kaleidoscope-LEDEffect-FunctionalColor/src/Kaleidoscope-LEDEffect-FunctionalColor.cpp:17: undefined reference to `kaleidoscope::LEDFunctionalColor::colorMapDefault::shift'
collect2: error: ld returned 1 exit status
exit status 1
Error compiling for board Keyboardio Model 01.
```

**Comment by jdlien** [comment]
Do I possibly have the plugin files in the wrong location? Does that even matter? It seems like the compiler is finding them -- or finding something with that name, at least. This is so weird.

**Comment by noseglasses** [comment]
How do you build it (command line/IDE), which platform do you use? Do things work if you comment FunctionalColors related stuff? Do you have any other (possibly dated) copy of the plugin in places where it could be picked up by the Arduino build system.
If all else fails, you could ask the community for help. Maybe someone else ran into the same or a similar problem, possibly with another plugin.

**Comment by noseglasses** [comment]
I just did another clean build with the latest Keyboardio-Bundle and FunctionalColor. On Ubuntu 18.04 everything builds flawlessly. It is already strange that in your build the compiler reports undefined references in a place where absolutely no reference is taken as everything (here the constexpr color definitions) is passed by value.
The reported line 17 is still more than suspicious as there is absolutely no relation to any color values in this line.

**Comment by noseglasses** [comment]
There are other possibly related reports about persisting build errors in issues https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/issues/4

As this discussion takes part in a closed PR, it is hard to find by others. @jdlien, could you possibly move the part of this thread starting from https://github.com/jdlien/Kaleidoscope-LEDEffect-FunctionalColor/pull/5#commitcomment-32406023 to a separate issue?

**Comment by lgnakano** [comment]
This is the same problem I have.  I was able to compile using the changes from ycros, and I am using that for now. Thanks for the help.

**Comment by noseglasses** [comment]
> This  is the same problem I have.  I was able to compile using the changes  from ycros, and I am using that for now. Thanks for the help.

While this apparently fixes things, I am pretty sure that it unecessarily blows up your firmware binary. That's no problem it your binary is small anyway but could be crucial for others.

@jdlien and @lgnakano, could you make your sketch available to me so I can try to find out what's the actual problem. In theory it should not be necessary to instanciate static constexpr globals, like you had to to solve your problem. I have a suspicion what is causing this but I need some basis to test my hypothesis. Unfortunately, my sketch builds fine. Also I would be interested in your Arduino version an build Platform.

**Comment by lgnakano** [comment]
I was having compilation errors with the simple example. I commented all of the funColors in examples/fc_example_simple.ino to make sure the compilation error I got was not from my side. Once I got that working with the changes from ycros, I added my macros and everything else on the layout.

**Comment by lgnakano** [comment]
I will report on which version of arduino I have once I can check. It is running on Windows 7 64-bit.

**Comment by noseglasses** [comment]
@Ignakano, do I understand that correctly. For you building the simple example without modifications fails?

**Comment by lgnakano** [comment]
Yes, that is correct.  I am using Arduino 1.8.8 in Windows 7 (sorry for the delay, real life got in the way).

**Pull request merged**